Android app

clemens-msupply opened the pull request:
This adds:
- Android project folder
- Build steps to build the remote-server shared lib

I didn't commit the changed yarn.lock yet...

mark-prins commented on README.md:
```diff
@@ -95,3 +95,80 @@ You can also specify multiple namespaces when using the hook:
+
+The Android app needs the remote server build as a shared lib.
+
+To build the shared lib the cross compiler needs to be build:
```
:nit (sorry! this is a very minor grammatical request)
```suggestion
To build the shared lib the cross compiler needs to be built:
```

mark-prins commented on README.md:
```diff
@@ -95,3 +95,80 @@ You can also specify multiple namespaces when using the hook:
+
+Note, unfortunately the remote server source directory can't be linked into the rust directory since otherwise `cargo` will not find the NDK directory.
+
+Unsolved issued that will hopefully be fixed at some point:
```
another minor :nit

```suggestion
Unsolved issues that will hopefully be fixed at some point:
```

mark-prins commented on README.md:
```diff
@@ -95,3 +95,80 @@ You can also specify multiple namespaces when using the hook:
+
+### Run the Android app
+
+In `packages/host/public/config.js` change `API_URL` to `API_URL: 'http://localhost:8082/graphql'` to use the remotes server running on Android.
```
```suggestion
In `packages/host/public/config.js` change `API_URL` to `API_URL: 'http://localhost:8082/graphql'` to use the remote server running on Android.
```

mark-prins reviewed (commented):
🎉  super exciting to see the app! thanks for getting it together 😁 

have run out of time today to review properly sorry! am running into issues compiling, so cannot comment helpfully yet. have just noted a couple of very minor wording :nits in the readme.

general thought though.. I would prefer this to be in a different repo. This repo is for the web app, and adding top level folders for the rust server, android & dist.. it breaks the pattern we have for organising files and the android app is a separate concern to the web app. Not sure if it is possible / easy to split out? 

Ideally the remote server and web app would be in their own respective repos, and the android app would be elsewhere, and reference the others. You've said that linking to the rust repo is an issue.. maybe we could organise the repos differently 🤔

clemens-msupply commented:
thanks fixed the spelling mistakes.

Yes I thought about having it in a separate repo. Think the capacitor.config.ts and the capacitor dependencies needs to be in the same npm repo though (think that would be the same for cordova). We could move the android folder into a different repo if needed.
On the Rust side I think its actually better to move the Android NDK generation and build process into the remote server repo. This would also make it possible to just link the repo into the frontend repo.

Another issue is that we probably want to have Android native code like bluetooth that we want to access from the web app. What I also like to do is to generate a https cert for the local remote server on first startup and then use it in the web app to have a secure connection on the device. So I need to communicate with the native app to get the public cert...
Not sure if we can easily split into a pure web and mobile version, e.g. build a version that support bluetooth and another version that is the pure web version...

No hurry to merge the PR, probably better to find a good solution first...

clemens-msupply commented:
I simplified the build process quite a bit:
- The NDK toolchain doesn't need to be build manually anymore but the one from the Android ndk is used now
- The makefile that builds the remote server android lib is in the remote server repo now (android-java-bindings
branch)
- The rust folder has been removed and it is assumed that the remote-server is located at `../remote-server`
- I added a `npm run android:build-remote-server` script to build the remote server libs and copy them to the android folder

Think it would be possible to move the Android dir into its own repo but I would advice against it. You probably want to make changes in the android folder and the FE folder at the same time and its much easier to have it in a single repo then.

mark-prins commented on README.md:
```diff
@@ -95,3 +95,68 @@ You can also specify multiple namespaces when using the hook:
+
+Unsolved issue that will hopefully be fixed at some point:
+
+1. in `../remote-server/Cargo.toml` add the "vendored" feature to openssl by changing the line:
```
```suggestion
1. in `../remote-server/server/Cargo.toml` add the "vendored" feature to openssl by changing the line:
```

clemens-msupply replied:
something went wrong in the last merge, this is not needed anymore; the feature is added automatically in remote-server when compiling the android target

mark-prins commented on README.md:
````diff
@@ -95,3 +95,68 @@ You can also specify multiple namespaces when using the hook:
+To build the remote server lib and copy it to the Android project run:
+
+```
+npm run android:build-remote_server
````
We're using yarn elsewhere in preference to npm, so it might be best to avoid confusion
```suggestion
yarn android:build-remote_server
```

mark-prins commented:
Building gives the error below, and I had to do what I was told (`rustup target add aarch64-linux-android`) to continue.
..which gives the next missing dependency: `aarch64-linux-android26-clang`.

Have NDK installed and the linker and compiler are found - am using the prebuilt version in toolchains. Will need to investigate further.

```
error[E0463]: can't find crate for `core`
  |
  = note: the `aarch64-linux-android` target may not be installed
  = help: consider downloading the target with `rustup target add aarch64-linux-android`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `cfg-if` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
make: *** [aarch64-linux-android] Error 101
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

mark-prins reviewed (commented):
just a comment at this stage.. still unable to compile.

joshxg commented:
👋 

Hit the same point as Mark .. had a little investigation but also need to dig deeper.

I did also want to look at https://github.com/cross-rs/cross and see if we could use that to cross compile which hopefully would take some of the annoyance out of it?

clemens-msupply commented:
Could you please check if you have the `aarch64-linux-android26-clang` bin in:

`$(NDK_HOME)/toolchains/llvm/prebuilt/$(HOST_TAG)/bin`

And if yes, could please let me know what the "HOST_TAG" directory is? I might get this directory wrong on darwin since I only tested it on linux...

mark-prins commented:
> Could you please check if you have the aarch64-linux-android26-clang bin in:

yes, it does exist. The `HOST_TAG` is `darwin-x86_64` for me.

I've changed the `$NDK_HOME` around and have a different error, again with sqlite3. This time

`sqlite3/sqlite3.c:14230:10: fatal error: 'stdio.h' file not found` Can have a look for the headers..

mark-prins reviewed (approved):
🎉  Woooo!!

or, in other words, that looks good. thanks. the apk builds and the app runs on my old lenovo 😁

mark-prins commented:
Confession time! I've made a few changes.. hope that's ok.
Main thing is moving the `android` directory to be under `packages`, as this makes thing neater 😁 
I've added some things into the readme, documenting the stumbling blocks for me on macOS. Have also added a script to build the apk, that makes life easier for me

mark-prins commented:
Also - while the app does run, there is an issue with the API ( app shows 'database error' ).
Would also like to change the splash screen and icon 😉 

Possibly: https://www.joshmorony.com/adding-icons-splash-screens-launch-images-to-capacitor-projects/
Or https://capacitorjs.com/docs/v2/guides/splash-screens-and-icons#creating-splash-screens-and-icons

clemens-msupply commented:
I forgot about the DB issue. What I did is to copy an initialized sqlite db to the right folder on the device. Shouldn't be a problem to do the db migration at first startup, on my TODO list...

mark-prins commented:
ah, thanks. I thought it might be the migration - but didn't see anything in the log output.
Realised this morning that the path in the script is now wrong 🙄  Will fix that up as an offroad
